src/chainstate/stacks/auth.rs

```diff
@@ -1488,39 +1483,24 @@ mod test {
         ];
 
         // we can serialize the invalid p2wpkh uncompressed condition, but we can't deserialize it
-        assert_eq!(bad_p2wpkh_uncompressed.serialize(), bad_p2wpkh_uncompressed_bytes);
+        let mut actual_bytes = vec![];
+        bad_p2wpkh_uncompressed.consensus_serialize(&mut actual_bytes).unwrap();
+        assert_eq!(actual_bytes, bad_p2wpkh_uncompressed_bytes);
         
         // we can serialize the invalid p2wsh uncompressed condition, but we can't deserialize it
-        assert_eq!(bad_p2wsh_uncompressed.serialize(), bad_p2wsh_uncompressed_bytes);
-
-        let mut index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_public_key_count_bytes, &mut index, bad_public_key_count_bytes.len() as u32).is_err());
-        assert_eq!(index, 0);
-        
-        let mut index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_public_key_count_bytes_2, &mut index, bad_public_key_count_bytes_2.len() as u32).is_err());
-        assert_eq!(index, 0);
-
-        index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_hash_mode_bytes, &mut index, bad_hash_mode_bytes.len() as u32).is_err());
-        assert_eq!(index, 0);
-
-        index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_hash_mode_multisig_bytes, &mut index, bad_hash_mode_multisig_bytes.len() as u32).is_err());
-        assert_eq!(index, 0);
-        
-        index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_p2wpkh_uncompressed_bytes, &mut index, bad_p2wpkh_uncompressed_bytes.len() as u32).is_err());
-        assert_eq!(index, 0);
-        
-        index = 0;
-        assert!(TransactionSpendingCondition::deserialize(&bad_p2wsh_uncompressed_bytes, &mut index, bad_p2wsh_uncompressed_bytes.len() as u32).is_err());
-        assert_eq!(index, 0);
+        let mut actual_bytes = vec![];
+        bad_p2wsh_uncompressed.consensus_serialize(&mut actual_bytes).unwrap();
+        assert_eq!(actual_bytes, bad_p2wsh_uncompressed_bytes);
+
+        assert!(TransactionSpendingCondition::consensus_deserialize(&mut io::Cursor::new(&bad_public_key_count_bytes)).is_err());
```

Comment:
Since `&[u8]` implements Read, you shouldn't need to do any wrapping with a cursor -- almost all of the `io::Cursor::new` instances could be `&mut bad_public_key_count_bytes.as_slice()`,

Reply:
No can do here.  `Read` on a `&[u8]` internally _mutates_ the `&[u8]` by shrinking it to just the unread part.  I'm using `io::Cursor` to avoid this.

Reply:
I believe it just mutates the reference, though, not the array itself.

Reply:
https://doc.rust-lang.org/src/std/io/impls.rs.html#211-275

Reply:
Right, the types are kind of funky -- that's a `&mut &[u8]`, which can mutate the reference, but not the array.

Reply:
Ah.  Will fix.

Reply:
Should be fixed now.